### PR TITLE
feat: new pill input type

### PR DESCRIPTION
**Pull request by kanikabansal-juspay**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Addition of a new input type which takes multiple input values convert them to array of pills.

https://github.com/user-attachments/assets/7ee5e12a-ce34-4653-9806-cc320ce8b605

## Motivation and Context

Requirement in other task and  provides a more intuitive way to display and manage multiple values in the UI.

## How did you test it?

Locally.

## Where to test it?

- [x] INTEG
- [x] SANDBOX
- [ ] PROD

## Checklist

- [x] I ran `npm run re:build`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: src/components/form/PillInput.res [hunk 1]
```diff
@@ -0,0 +1,147 @@
+@react.component
+let make = (~name, ~initialItems: array<string>=[], ~placeholder, ~duplicateCheck=true) => {
+  let form = ReactFinalForm.useForm()
+  let (items, setItems) = React.useState(() => initialItems)
+  let (inputValue, setInputValue) = React.useState(() => "")
+  //to get edited input val
+  let (editInput, setEditInput) = React.useState(() => "")
```
Comment (PritishBudhiraja) [inline src/components/form/PillInput.res:7]:
```suggestion
  let (editInput, setEditInput) = React.useState(_ => "")
```
Comment can be removed. Do for the rest too for _=>"".

File: src/components/form/PillInput.res [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+  let handleEditChange = e => {
+    let value = ReactEvent.Form.target(e)["value"]
+    setEditInput(_ => value)
+    setError(_ => "")
+  }
+
+  let addItem = item => {
+    let trimmedItem = String.trim(item)
+    if trimmedItem == "" {
```
Comment (PritishBudhiraja) [inline src/components/form/PillInput.res:26]:
```suggestion
    if trimmedItem->LogicUtils.isEmptyString {
```

File: src/components/form/PillInput.res [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        className="outline-none p-2 flex-grow"
+        name
+      />
+    </div>
+    {error != ""
+      ? <div className="flex gap-1">
+          <Icon name="exclamation-circle" size=14 className="!text-red-500 mt-1" />
+          <p className="text-red-700 text-sm mt-1"> {React.string(error)} </p>
+        </div>
+      : React.null}
```
Comment (PritishBudhiraja) [inline src/components/form/PillInput.res:145]: Use RenderIf and isNonStringEmpty

File: src/components/form/PillInput.res [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                }
+              }}
+              className="rounded-md p-1 text-sm flex-grow"
+            />
+            <button
+              onClick={_ => removeItem(item)} className="ml-2 hover:bg-gray-300 rounded-md p-1">
+              <Icon name="cross-outline" size=14 />
+            </button>
+          </div>
+        | _ =>
```
Comment (PritishBudhiraja) [inline src/components/form/PillInput.res:115]: Add explicitly the None case

File: src/components/form/PillInput.res [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+              onBlur={_ => saveItem(item)}
+              onInput=handleEditChange
+              onKeyDown={ev => {
+                open ReactEvent.Keyboard
+                let key = ev->key
+                let keyCode = ev->keyCode
+                if key === "Enter" || keyCode === 13 {
+                  ReactEvent.Keyboard.preventDefault(ev)
+                  saveItem(item)
+                }
```
Comment (PritishBudhiraja) [inline src/components/form/PillInput.res:106]: Don't open use directly I feel so only one line is needed

File: src/components/form/PillInput.res [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  }
+
+  let handleKeyDown = e => {
+    open ReactEvent.Keyboard
+    let key = e->key
+    let keyCode = e->keyCode
+    if key === "Enter" || keyCode === 13 {
+      ReactEvent.Keyboard.preventDefault(e)
+      addItem(inputValue)
```
Comment (PritishBudhiraja) [inline src/components/form/PillInput.res:78]: Same don't open

File: src/components/form/PillInput.res [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    let trimmedItem = String.trim(item)
+    if trimmedItem == "" {
+      setInputValue(_ => "")
+      setError(_ => "")
+    }
+    if duplicateCheck && Array.some(items, existingItem => existingItem == trimmedItem) {
+      setError(_ => "Email already exists")
+    } else if !CommonAuthUtils.isValidEmail(trimmedItem) {
+      setItems(prev => Array.concat(prev, [trimmedItem]))
+      form.change(name, items->Array.concat([item])->Identity.genericTypeToJson)
```
Comment (PritishBudhiraja) [inline src/components/form/PillInput.res:34]:
```suggestion
      form.change(name, [...items, ...item])->Identity.genericTypeToJson)
```

I think names can be given a bit diff

**Review by PritishBudhiraja — COMMENTED** [review]
Can you add a video how it will look. Rest looks fine

File: src/components/form/PillInput.res [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    form.change(name, items->Array.filter(ele => ele !== itemToRemove)->Identity.genericTypeToJson)
+    setItems(prev => prev->Array.filter(item => item != itemToRemove))
+  }
+
+  let saveItem = itemToSave => {
+    let trimmedEditInput = String.trim(editInput)
+    if (
+      duplicateCheck &&
+      trimmedEditInput != itemToSave &&
+      Array.some(items, existingItem => existingItem == trimmedEditInput)
```
Comment (sagarnaikjuspay) [inline src/components/form/PillInput.res:69]: can u store this in some external variable and use in if condition

File: src/components/form/PillInput.res [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+      ->Array.mapWithIndex((item, i) =>
+        switch editingItem {
+        | Some(selected) if selected == item =>
+          <div key={Int.toString(i)} className="flex flex-wrap gap-1 p-1 border text-sm rounded-md">
+            <input
+              type_="text"
+              value={editInput}
+              onBlur={_ => saveItem(item)}
+              onInput=handleEditChange
+              onKeyDown={ev => {
```
Comment (sagarnaikjuspay) [inline src/components/form/PillInput.res:118]: this also can be written in external handleKeyDown function and used here

File: src/components/form/PillInput.res [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+                let key = ev->ReactEvent.Keyboard.key
+                let keyCode = ev->ReactEvent.Keyboard.keyCode
+                if key === "Enter" || keyCode === 13 {
```
Comment (sagarnaikjuspay) [inline src/components/form/PillInput.res:121]: and "Enter" and 13 these constants can be stored in the variable on used , its used another key handling function as well

File: src/components/form/PillInput.res [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+              <div className="flex items-center gap-2">
+                <img alt="user_icon" src={`/icons/user_icon.svg`} className="h-6 w-6" />
+                <span className="font-medium"> {React.string(suggestion->Option.getOr(""))} </span>
+              </div>
+            </div>
+          </div>
+        </RenderIf>
+      </div>
+    </div>
+    <RenderIf condition={!(error->LogicUtils.isEmptyString)}>
```
Comment (sagarnaikjuspay) [inline src/components/form/PillInput.res:172]: `isNonEmptyString` can be used directly, and logicUtils can be opend at once at once if its used inmultiple places

Comment (kanikabansal-juspay) [reply src/components/form/PillInput.res:172]: It's used in just 2 places.

File: src/components/form/PillInput.res (hunk 4 above)
Comment (PritishBudhiraja) [reply src/components/form/PillInput.res:133]: It's either Some or None. You won't be having more. Correct?

Comment (kanikabansal-juspay) [reply src/components/form/PillInput.res]: Yes Made changes in the switch conditions.

**Review by PritishBudhiraja — DISMISSED** [review]

**Review by PritishBudhiraja — APPROVED** [review]
LGTM

**Review by sagarnaikjuspay — APPROVED** [review]

**Pull request merged**